File: packages/js/core/src/genai/top_p.ts
import { Tensor } from '../index.js';
import { LogitProcessor } from './logit_processors.js';

export class TopPLogitProcessor implements LogitProcessor {
  constructor(private topP: number) {
    if (topP <= 0 || topP > 1.0) {
      throw new Error('topP must be in (0, 1].');
    }
  }

  process(inputIds: number[], logits: Tensor): Tensor {
    if (this.topP >= 1.0 || !(logits.data instanceof Float32Array)) {
      return logits;
    }

    const vocabSize = logits.shape[logits.shape.length - 1] as number;
    const offset = logits.data.length - vocabSize;

    // Apply softmax first to get probabilities
    const maxLogit = Math.max(...Array.from(logits.data.subarray(offset)));
    const probs = new Float32Array(vocabSize);
    let sumProbs = 0;

    for (let i = 0; i < vocabSize; i++) {
      const data = logits.data as Float32Array;
      probs[i] = Math.exp(data[offset + i]! - maxLogit);
      sumProbs += probs[i]!;
    }

    for (let i = 0; i < vocabSize; i++) {
      probs[i] = probs[i]! / sumProbs;
    }

    const vals: { val: number; idx: number }[] = [];
    for (let i = 0; i < vocabSize; i++) {
      vals.push({ val: probs[i]!, idx: i });
    }

    vals.sort((a, b) => b.val - a.val);

    let cumulativeProb = 0;
    let thresholdIdx = vocabSize - 1;

    for (let i = 0; i < vocabSize; i++) {
      cumulativeProb += vals[i]!.val;
      if (cumulativeProb > this.topP) {
        thresholdIdx = i;
        break;
      }
    }

    // We keep at least one token
    const indicesToRemove = vals.slice(thresholdIdx + 1).map((x) => x.idx);
    if (indicesToRemove.length > 0) {
      const newData = new Float32Array(logits.data);
      for (const idx of indicesToRemove) {
        newData[offset + idx] = -Infinity;
      }
      return new Tensor(
        logits.name,
        logits.shape,
        logits.dtype,
        logits.isInitializer,
        logits.requiresGrad,
        newData,
      );
    }

    return logits;
  }
}

File: packages/python/onnx9000-core/src/onnx9000/genai/top_p.py
"""Provide functionality for this module."""

import math

from ..core.ir import Tensor
from .logit_processors import LogitProcessor


class TopPLogitProcessor(LogitProcessor):
    """Filters logits based on top-P (Nucleus) sampling."""

    def __init__(self, top_p: float):
        """Initialize the instance."""
        if top_p <= 0.0 or top_p > 1.0:
            raise ValueError("top_p must be strictly positive and <= 1.0")
        self.top_p = top_p

    def __call__(self, input_ids: list[int], logits: Tensor) -> Tensor:
        """Execute the callable."""
        if self.top_p >= 1.0 or logits.data is None:
            return logits

        import struct

        itemsize = logits.dtype.itemsize if hasattr(logits.dtype, "itemsize") else 4
        vocab_size = logits.shape[-1] if logits.shape else len(logits.data) // itemsize

        offset = len(logits.data) - vocab_size * itemsize

        vals = []
        for i in range(vocab_size):
            start = offset + i * itemsize
            if start >= 0 and start + itemsize <= len(logits.data):
                val = struct.unpack("<f", logits.data[start : start + itemsize])[0]
                vals.append((val, i))

        # Softmax
        max_val = max(v[0] for v in vals) if vals else 0
        exp_vals = [math.exp(v[0] - max_val) for v in vals]
        sum_exp = sum(exp_vals)
        probs = [e / sum_exp for e in exp_vals]

        # Sort by probability
        sorted_probs = sorted(zip(probs, [v[1] for v in vals]), key=lambda x: x[0], reverse=True)

        cumulative_prob = 0.0
        threshold_idx = len(sorted_probs) - 1

        for i, (prob, idx) in enumerate(sorted_probs):
            cumulative_prob += prob
            if cumulative_prob > self.top_p:
                threshold_idx = i
                break

        # Indices to mask
        indices_to_remove = [idx for _, idx in sorted_probs[threshold_idx + 1 :]]

        if indices_to_remove:
            new_data = bytearray(logits.data)
            neg_inf = struct.pack("<f", float("-inf"))
            for idx in indices_to_remove:
                start = offset + idx * itemsize
                new_data[start : start + itemsize] = neg_inf
            return Tensor(name=logits.name, shape=logits.shape, dtype=logits.dtype, data=new_data)

        return logits
